File: packages/tonic-core/src/hydration/conflictHunkExcerpt.ts
import * as fs from "node:fs";
import * as path from "node:path";

import { parseGitConflictsWithDiagnostics } from "../gitConflictParser";
import type { ConflictContextArtifactV1, ConflictRegionV1 } from "./conflictScan";

export const CONFLICT_HUNK_MAX_LINES_PER_SIDE = 40;
export const CONFLICT_HUNK_MAX_TOTAL_CHARS = 32000;

export type ConflictHunkExcerptRegionV1 = {
  region_id?: string;
  path: string;
  ours_excerpt: string;
  theirs_excerpt: string;
  truncated?: boolean;
};

export type ConflictHunkExcerptsArtifactV1 = {
  schema: "tonic-conflict-hunk-excerpts";
  version: "1";
  regions: ConflictHunkExcerptRegionV1[];
};

function truncateLines(lines: string[], maxLines: number): { text: string; truncated: boolean } {
  if (lines.length <= maxLines) {
    return { text: lines.join("\n"), truncated: false };
  }
  const slice = lines.slice(0, maxLines);
  return {
    text: slice.join("\n") + "\n[… truncated …]",
    truncated: true,
  };
}

/**
 * Build bounded ours/theirs text per conflict region for LLM context.
 * Respects `TONIC_CONFLICT_EXCERPT=0` to disable (returns empty regions array).
 */
export function buildConflictHunkExcerpts(
  repoRoot: string,
  conflictArt: ConflictContextArtifactV1,
  env: NodeJS.ProcessEnv,
): ConflictHunkExcerptsArtifactV1 {
  const disabled = (env.TONIC_CONFLICT_EXCERPT ?? "").trim() === "0";
  if (disabled || conflictArt.conflict_regions.length === 0) {
    return { schema: "tonic-conflict-hunk-excerpts", version: "1", regions: [] };
  }

  const maxLines = Math.max(
    1,
    parseInt(env.TONIC_CONFLICT_EXCERPT_MAX_LINES_PER_SIDE ?? "", 10) || CONFLICT_HUNK_MAX_LINES_PER_SIDE,
  );
  const maxTotal = Math.max(
    1024,
    parseInt(env.TONIC_CONFLICT_EXCERPT_MAX_TOTAL_CHARS ?? "", 10) || CONFLICT_HUNK_MAX_TOTAL_CHARS,
  );

  const regions: ConflictHunkExcerptRegionV1[] = [];
  let totalChars = 0;

  for (const r of conflictArt.conflict_regions) {
    if (totalChars >= maxTotal) {
      break;
    }
    const abs = path.join(repoRoot, r.path.replace(/\//g, path.sep));
    let text: string;
    try {
      text = fs.readFileSync(abs, "utf8");
    } catch {
      continue;
    }
    const { blocks } = parseGitConflictsWithDiagnostics(text);
    const start0 = r.start_line - 1;
    const end0 = r.end_line - 1;
    let matched = false;
    for (const b of blocks) {
      if (b.startLine !== start0 || b.endLine !== end0) {
        continue;
      }
      matched = true;
      const oursLines = b.segments[0]?.lines ?? [];
      const theirsLines = b.segments[1]?.lines ?? [];
      const o = truncateLines(oursLines, maxLines);
      const t = truncateLines(theirsLines, maxLines);
      const row: ConflictHunkExcerptRegionV1 = {
        region_id: r.region_id,
        path: r.path,
        ours_excerpt: o.text,
        theirs_excerpt: t.text,
        truncated: o.truncated || t.truncated,
      };
      const rowChars = row.ours_excerpt.length + row.theirs_excerpt.length + 64;
      if (totalChars + rowChars > maxTotal) {
        row.ours_excerpt = row.ours_excerpt.slice(0, Math.max(0, maxTotal - totalChars - 200)) + "\n[… truncated …]";
        row.theirs_excerpt = "";
        row.truncated = true;
      }
      totalChars += row.ours_excerpt.length + row.theirs_excerpt.length + 64;
      regions.push(row);
      break;
    }
    if (!matched) {
      regions.push({
        region_id: r.region_id,
        path: r.path,
        ours_excerpt: "",
        theirs_excerpt: "",
        truncated: false,
      });
    }
  }

  return { schema: "tonic-conflict-hunk-excerpts", version: "1", regions };
}

export function conflictHunkExcerptsToPromptJson(art: ConflictHunkExcerptsArtifactV1): string {
  return JSON.stringify(art.regions, null, 2);
}

export function mergeBranchHintsFromRegions(regions: ConflictRegionV1[]): string {
  const labels: string[] = [];
  for (const r of regions) {
    if (r.ours_label) {
      labels.push(`ours=${r.ours_label}`);
    }
    if (r.theirs_label) {
      labels.push(`theirs=${r.theirs_label}`);
    }
  }
  if (labels.length === 0) {
    return "";
  }
  return [...new Set(labels)].join("; ");
}

File: merge-tonic-lib/tonic/hydration/conflict_hunk_excerpt.py
"""Bounded ours/theirs conflict text for LLM prompts (parity with conflictHunkExcerpt.ts)."""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any

from tonic.git_conflict_parser import parse_git_conflicts

CONFLICT_HUNK_MAX_LINES_PER_SIDE = 40
CONFLICT_HUNK_MAX_TOTAL_CHARS = 32000


def _truncate_lines(lines: list[str], max_lines: int) -> tuple[str, bool]:
    if len(lines) <= max_lines:
        return "\n".join(lines), False
    slice_ = lines[:max_lines]
    return "\n".join(slice_) + "\n[… truncated …]", True


def build_conflict_hunk_excerpts(
    repo_root: str,
    conflict_art: dict[str, Any],
    env: dict[str, str],
) -> dict[str, Any]:
    disabled = (env.get("TONIC_CONFLICT_EXCERPT") or "").strip() == "0"
    regions_in = list(conflict_art.get("conflict_regions") or [])
    if disabled or not regions_in:
        return {"schema": "tonic-conflict-hunk-excerpts", "version": "1", "regions": []}

    max_lines = max(
        1,
        int((env.get("TONIC_CONFLICT_EXCERPT_MAX_LINES_PER_SIDE") or "").strip() or "0") or CONFLICT_HUNK_MAX_LINES_PER_SIDE,
    )
    max_total = max(
        1024,
        int((env.get("TONIC_CONFLICT_EXCERPT_MAX_TOTAL_CHARS") or "").strip() or "0") or CONFLICT_HUNK_MAX_TOTAL_CHARS,
    )

    root = Path(repo_root)
    out_regions: list[dict[str, Any]] = []
    total_chars = 0

    for r in regions_in:
        if total_chars >= max_total:
            break
        rel = str(r.get("path") or "").replace("\\", "/")
        parts = [x for x in rel.split("/") if x]
        if not parts:
            continue
        abs_p = root.joinpath(*parts)
        try:
            text = abs_p.read_text(encoding="utf-8", errors="replace")
        except OSError:
            continue
        blocks = parse_git_conflicts(text)
        start0 = int(r.get("start_line", 0)) - 1
        end0 = int(r.get("end_line", 0)) - 1
        matched = False
        for b in blocks:
            if b.start_line != start0 or b.end_line != end0:
                continue
            matched = True
            ours_lines = b.segments[0].lines if b.segments else []
            theirs_lines = b.segments[1].lines if len(b.segments) > 1 else []
            o_t, o_tr = _truncate_lines(ours_lines, max_lines)
            t_t, t_tr = _truncate_lines(theirs_lines, max_lines)
            row: dict[str, Any] = {
                "region_id": r.get("region_id"),
                "path": str(r.get("path") or "").replace("\\", "/"),
                "ours_excerpt": o_t,
                "theirs_excerpt": t_t,
                "truncated": o_tr or t_tr,
            }
            row_chars = len(row["ours_excerpt"]) + len(row["theirs_excerpt"]) + 64
            if total_chars + row_chars > max_total:
                row["ours_excerpt"] = row["ours_excerpt"][: max(0, max_total - total_chars - 200)] + "\n[… truncated …]"
                row["theirs_excerpt"] = ""
                row["truncated"] = True
            total_chars += len(row["ours_excerpt"]) + len(row["theirs_excerpt"]) + 64
            out_regions.append(row)
            break
        if not matched:
            out_regions.append(
                {
                    "region_id": r.get("region_id"),
                    "path": str(r.get("path") or "").replace("\\", "/"),
                    "ours_excerpt": "",
                    "theirs_excerpt": "",
                    "truncated": False,
                }
            )

    return {"schema": "tonic-conflict-hunk-excerpts", "version": "1", "regions": out_regions}


def conflict_hunk_excerpts_to_prompt_json(art: dict[str, Any]) -> str:
    return json.dumps(art.get("regions") or [], indent=2)


def merge_branch_hints_from_regions(regions: list[dict[str, Any]]) -> str:
    labels: list[str] = []
    for r in regions:
        ol = r.get("ours_label")
        tl = r.get("theirs_label")
        if ol:
            labels.append(f"ours={ol}")
        if tl:
            labels.append(f"theirs={tl}")
    if not labels:
        return ""
    return "; ".join(dict.fromkeys(labels))
